### PR TITLE
Implement p-norm definitions

### PR DESCRIPTION
**Pull request by polvalente**
This PR aims to implement the 0-norm, 1-norm and 2-norm cited in #157 via the `p_norm/2` function

### CODE REVIEW

**Comment by seanmor5** [comment]
Hey thanks for the PR!

A few notes before we can merge:

- We generally try to adhere to NumPy semantics where it makes sense, I would take a look at [np.linalg.norm](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html). I think in this case it's okay to adhere to their semantics. I would use `:ord`, `:keep_axes`, and `:axes` as optional keywords in `opts`. Sum supports both `keep_axes` and `axes` as well so you should be able to just plug those in there. You'll also want to take a look at the usage of `assert_keys`.

- You should assert upfront that the input `t` is a tensor, it will make the call stack clearer if somebody passes invalid input to this function, rather than it getting caught down the line:

```elixir
def p_norm(tensor, opts \\ []) do
  %T{type: type} = tensor = tensor!(tensor)
   ...
```
- To go along with the discussion of numpy semantics, you'll want to do some shape checks on the inputs. I believe numpy only supports 1-D/2-D inputs depending on `:ord`, so you'll need to do that validation, unless you think there's a reason not to.

- Be careful with automatic promotion to `f64`, it's not supported on all accelerators (TPU), and it's generally really slow on both CPU/GPU. Instead I would do something like:

```
def p_norm(tensor, opts \\ []) do
  %T{type: type} = tensor = tensor!(tensor)
  output_type = Nx.Type.to_floating(type)
  ...
  |> Nx.as_type(output_type)
end
```

That will ensure you're not aggressively upcasting to `f64`.

- This is a good implementation, and will make p_norm available automatically for EXLA and other compilers. If you're up for it, you can also write some tests for the EXLA compiler in `exla/test/exla/defn_api_test.exs`. There are lots of examples there. The tests should pass without any additions to EXLA.

**Comment by polvalente** [comment]
@seanmor5 sure, I'll try to adapt to the NumPy interface :)

At first, I don't see a reason to restrict dimensions, since my implementation basically returns a scalar metric. However, upon defining `axes` and such this might change. I'll keep this in mind!

Thanks for the tip on `f64`. My idea was to make all norms return the same datatype, but returning the same type as the input for `0-norm` makes sense!

**Comment by AlexJuca** [comment]
@polvalente I see you have a PR for this already. Looks like I will leave it to you to handle. It's a great start.

**Comment by polvalente** [comment]
@seanmor5 I think I've done everything you asked! I had some trouble supporting the `axes` and `keep_axes` options, but I think I got it. The only problem that I couldn't solve is the implementation for `norm(t, -2)` when `t` is a 2-D tensor, because the numpy docs say it returns the minimum singular value for the matrix.

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6773]: There is a function called `assert_keys!` that you should use to validate the inputs.

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6751]: Could we use a numerically stable implementation? https://timvieira.github.io/blog/post/2014/11/10/numerically-stable-p-norms/

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6665]: Please add a description of the supported options. Note that if `:fro` and `:nuc` are abbreviations, prefer to use full words instead. :)

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6773]: Also, those are default values, so this should be fine:

```suggestion
    axes_opts = Keyword.take(input_opts, [:axes, :keep_axes])
```

**Comment by josevalim** [comment]
Looks great, thank you! :heart: I have added some comments too! Btw, should we call this norm or p_norm?

**Comment by AlexJuca** [comment]
@josevalim @polvalente  I would suggest following the numpy semantics and have it be the more generic `norm()`.

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6773]: I'll take a further look into it

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6751]: That's a great suggestion! I'll try make the proper adjustments for this.
Do we have a function that yields the maximum value for each datatype?
If not, maybe we should leave this as a technical debt to avoid hardcoding any values. I'm thinking of something like `Nx.max_value(1.0)` yielding `1.0e300`

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6665]: Sure! I forgot to update the docs after adding said options. I'll improve naming as well.

**Comment by polvalente** [comment]
> @josevalim @polvalente I would suggest following the numpy semantics and have it be the more generic norm.

I'm with @AlexJuca. `p_norm` would imply only the base case of `power(sum(power(abs(x), p)), 1/p)`, whereas `norm` includes the special cases documented here: https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html

By the way, @josevalim is there a good way to draw tables in docs? Maybe I should also port the norm definitions table itself to our doc.

File: nx/lib/nx.ex
Comment (josevalim) [reply nx/lib/nx.ex:6751]: Yes, there is `Nx.Type.min_value_binary` which you can put into a tensor by calling `Nx.from_binary`.

**Comment by polvalente** [comment]
@seanmor5 @josevalim @AlexJuca I've made the latest changes with better docs. I only have some questions on how to deal with `ord: -2` for 2-D Tensors and with `ord: :nuc`, both of which depend on SVD to work.

**Comment by AlexJuca** [comment]
@polvalente both cases can always be deferred for later. You can raise saying that the operation is not supported yet. What you think?

**Comment by polvalente** [comment]
> @polvalente both cases can always be deferred for later. Seeing that they are dependent on SVD you can raise saying that the operation is not supported yet. What you think?

Sounds like a good plan!

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6849]: Do all of NumPy's cases return a float? If not, ignore this :)

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6873]: I think you can combine these two:

```elixir
defp do_p_norm(t, ord, rank, opts) when ord in [:inf, :neg_inf] and rank == 1 or rank == 2 do
  function = if ord == :inf, do: &reduce_max/1, else: &reduce_min/1
  reduction = if rank == 1, do: & &1, else: &sum(&1, [keep_axes: opts[:keep_axes], axes: [1]])

  t
  |> abs()
  |> reduction.()
  |> function.()
end
```

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6887]: I think you might be able to combine this with the case above as well :)

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6894]: I would be careful using `divide` here because it might cause unnecessary type promotion later on. Instead I would do:

```elixir
inv_ord = tensor(1 / ord, type: type)
```

Also, is it possible `ord` is 0? Maybe it was caught earlier, but if not, be careful of division by 0 :)

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6904]: You don't need the fetch here, the max value scalar will work fine in subsequent operations!

**Review by seanmor5 — COMMENTED** [review]
Hey @polvalente great work! I understand you need SVD to get this finished. For now, if you're up for it, you can try to implement a somewhat naive version of SVD and I can finish the rest for you on the EXLA side. If not, raising is fine for now and I will just add a note to the issue tracker.

**Comment by polvalente** [comment]
> Hey @polvalente great work! I understand you need SVD to get this finished. For now, if you're up for it, you can try to implement a somewhat naive version of SVD and I can finish the rest for you on the EXLA side. If not, raising is fine for now and I will just add a note to the issue tracker.

Sure! Do you want me to add the naive SVD in this PR, or should we merge this first and I add it in a subsequent one? I lean towards this second option because L2-norm could help with QR decomposition and other things like that.

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6904]: The intention here was to avoid having a [1]-shaped Tensor  instead of a pure scalar (altough they are almost the same) in some of the test results.

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6894]: There are special function clauses for `ord: 0`. The numpy docs specified it in a similar manner to what I had in the original draft: returns the count of non-zero elements in the tensor

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6873]: ooh this seems nice :)
I'll look into this!

**Comment by seanmor5** [comment]
@polvalente You can go ahead and get this finished up and worry about SVD later! I would raise on the unsupported cases for now then, and once this is merged I'll add a note to the issues tracker

File: exla/test/exla/defn_expr_test.exs
Comment (polvalente) [reply exla/test/exla/defn_expr_test.exs:2395]: Sure!

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6770]: Sounds good. Should I really keep the error cases as doctests/examples? Or maybe add them to nx_test.exs?

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6783]: no problems!

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6788]: I tend to use this approach so all defaults are specified in a single place. Do you feel the need for this change?

File: nx/lib/nx.ex
Comment (seanmor5) [reply nx/lib/nx.ex:6770]: I lean towards keeping at least 1 error case in the doctests, just for documentation purposes!

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6806]: I feared this would be an anti-pattern for Nx, since the Tensor is `inspect`ed as an opaque struct. Since this is ok, I'll refactor this as well!

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6849]: Some of them return 1-D tensors for 2-D inputs

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6885]: I ran into some problems using `type: type` because `type` can represent an integer type. I've refactored this line to `tensor(1/p)` instead

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex]: This one is possible, but I think it removes some readability because it's only defined for matrices

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex]: I've simplified this __a lot__, so we can disregard my comment :)

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6779]: Nitpick: don't end error messages with dot. :)

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6688]:
```suggestion
  ### Vector norms
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6690]: Make sure the examples are indented with 4 spaces from the heredoc beginning indentation, as that's required by Markdown. Otherwise they won't be highlighted as code snippets. TL;DR - indent all examples by two spaces. :)

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6792]:
```suggestion
      do: raise(ArgumentError, "expected 1-D or 2-D tensor, got tensor with shape #{inspect(s)}")
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6804]:
```suggestion
    do: raise(ArgumentError, ":frobenius expects a 2-D tensor, got a 1-D tensor")
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6815]:
```suggestion
    raise ArgumentError, "expected 1-D tensor for ord: 0, got a 2-D tensor"
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6801]: If we don't implement it yet, maybe we should simply not support this option. :)

Comment (josevalim) [inline nx/lib/nx.ex:6801]:
```suggestion
  defp do_p_norm(_t, :nuclear, _opts), do: raise("nuclear norm not implemented yet")
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6849]:
```suggestion
    raise ArgumentError, "invalid :ord for 2-D tensor, got: #{inspect(ord)}"
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6845]:
```suggestion
    raise "ord: -2 for 2-D tensor not implemented yet"
```

File: nx/lib/nx.ex
Comment (josevalim) [inline nx/lib/nx.ex:6796]: I think we can slightly cleanup the code by handling the major ord branches here:

```
case ord do
  :frobinius -> norm_frobinius(t, axes_opts)
  :nuclear -> norm_nuclear(t, axes_opts)
  _ when ord in [:inf, :neg_inf] -> norm_inf(t, ord, axes_opts)
  _ when is_integer(ord) -> norm_ord(t, ord, axes_opts)
  _ -> raise ArgumentError, "unknown ord #{inspect(ord)}"
end
```

**Review by josevalim — COMMENTED** [review]
I have added some minor nitpicks to the error messages to make them consistent. I have also made a proposal to reduce the branching inside do_p_norm. We are almost there I think, thank you!

File: nx/lib/nx.ex
Comment (seanmor5) [inline nx/lib/nx.ex:6859]: Sorry I lead you astray last time, this will promote to `f64` rather quickly which you want to avoid. It can be fixed on the user end, but in general it's better to make sure we're not doing it unnecessarily if possible. This is what you want instead: 

```suggestion
  defp norm_integer(%{type: type} = t, ord, opts) when is_integer(ord) do
    output_type = Nx.Type.to_floating(type)
    inv_ord = tensor(1 / ord, type: output_type)
```

**Review by seanmor5 — APPROVED** [review]
Just the 1 comment on type promotion, once that's fixed this looks good to me :)

File: nx/lib/nx.ex
Comment (polvalente) [reply nx/lib/nx.ex:6859]: No worries!

**Pull request merged**
